### ISSUE TITLE
A bugged Timeline/Challenge row in the Profile, duplicates itself.

### ISSUE THREAD

**Issue by eldarlrd** [freeCodeCamp/freeCodeCamp#47173]
### Describe the Issue

I just finished the "Build a Tribute Page" challenge, now when I go to check my timeline it shows the challenge doubled. Going back and forth from the Timeline page that has the bugged row, adds more duplicate "Build a Tribute Page" rows.

When I go to the challenge again and Run the code, it shows me double the percentage I should've received i.e. 60% instead of 40%.

### Affected Page

https://www.freecodecamp.org/eldarlrd

### Steps to Reproduce

1. Go to 'Timeline'
2. Click on 'Page 19'
3. Scroll down to 'Build a Tribute Page'

### Expected behavior

It shouldn't duplicate.

### System

- Device: Desktop PC, Smartphone
- OS: Arch Linux, Android
- Browser: Mozilla Firefox
- Version: 103.0.1

### Additional context

The issue seems to be server-side as it can be reproduced on a different device too.

**Comment by DylanDevelops** [comment]
Could you elaborate on your issue I'm a little confused what the error is? I see this: 

![image](https://user-images.githubusercontent.com/48571264/183235603-e99b3292-a370-467b-9251-4a462697b6bd.png)

**Comment by eldarlrd** [comment]
> Could you elaborate on your issue I'm a little confused what the error is? I see this:
> 
> ![image](https://user-images.githubusercontent.com/48571264/183235603-e99b3292-a370-467b-9251-4a462697b6bd.png)

There are two "Build a Tribute Page" rows when there only should be one. Also, try going back to Page 4 or forward to Page 6 and you'll see a bugged "Build a Tribute Page" on top of the list, out of place.

**Comment by DylanDevelops** [comment]
Oh I see it now. I'm unaware if normal contributors have the ability to change the way the profile code works so you might have to wait for a member of freecodecamp to come along and see this issue.

**Comment by ojeytonwilliams** [comment]
My PR, https://github.com/freeCodeCamp/freeCodeCamp/pull/45862, is probably to blame.  

I suspect what's happening is that multiple submit requests can end up reaching the api more or less simultaneously.  The api gets the user object from the db in one query, looks to see if the challenge has been completed and then sends another query to update the db. Two simultaneous requests will see the same user object and both add to the array of completed challenges.

I can see three possible solutions

1. Recreate the completeChallenges array on each submission (i.e. revert my PR)
2. Achieve eventual correctness by enforcing uniqueness if the challenge array is found to contain duplicates
3. Use a transaction

Option 1. puts more load on the db, but I'm not sure how significant it is.  Option 2. feels a bit sketchy. If a camper stops submitting challenges after their final project submission, their profile will never get fixed. Option 3. should work, but I'm not sure if it's possible with loopback.

In addition, the timeline gets messed up if there are duplicate entries, but that's less of an issue.

@raisedadead @naomi-lgbt I'm leaning towards option 1. what do you reckon?

**Comment by ojeytonwilliams** [comment]
Also, one other idea I considered was to remove any duplicates with `$pull` and then `$push` the updated submission.  Unfortunately, unless done as a transaction, this solves nothing. Simultaneous updates will `$pull` in parallel and `$push` in parallel, so effectively the `$pull` never happens.

**Comment by raisedadead** [comment]
In an ideal world, it should be a perfect timeline, but practically speaking, we do not even need it to be consistent on the DB. 

We can sort all the entries and filter out duplicates before sending the list downstream to the client. It should be a relatively quick fix vs. rejigging the whole thing and tripping over elsewhere because we do not have a consistent schema on the documents, to begin with.

**Comment by ojeytonwilliams** [comment]
I'm not worried about the timeline, exactly, I'm worried about what this means for everything that depends on the `completedChallenges` data.  If there are duplicates of a given project, sure, we could filter them out before we send anything to the client, but that means we have to do that every time want to access that array.

We introduced a load of bugs because we changed `files` to `challengeFiles` and missed a bunch of places where that mattered. I don't think we should do the same kind of thing for `completedChallenges`.

**Comment by raisedadead** [comment]
I see - I'll defer to you. If this means the heightened load on the cluster, then I am okay with it as long as Quincy signs off on the costing to convenience tradeoff.

**Comment by ojeytonwilliams** [comment]
Okay, cool, I'll check in with him.

**Comment by ojeytonwilliams** [comment]
As a reminder, because I'll likely need my memory jogging:

The issue is that concurrent submissions to the api will...

1. trigger two checks to see if the submitted challenge is not in the db
2. both find that it is not
3. push the same challenge onto completed challenges twice

Also, I missed a solution last time I considered this: migrate to a datastructure that handles concurrent submissions better.  An object

```js
{ 
 id1: { challengeData },
 id2: { moreChallengeData }
}
```

could do the same checks

1. trigger two checks to see if the submitted challenge is not in the db
2. both find that it is not
3. update the challenge object twice

In this case the second submission will win and there will not be any duplication.

**Comment by eldarlrd** [comment]
I just noticed that the issue has been resolved. The duplicates are gone, thank you.
